### PR TITLE
update Access-Control-Allow-Headers to enumerate rather than wildcard

### LINKED ISSUE #4045: Access-Control-Allow-Headers CORS bug
safari and firefox don't (yet) support wildcards on Access-Control-Allow-Headers, despite this being adopted by whatwg in 2016 😡 :

firefox: https://bugzilla.mozilla.org/show_bug.cgi?id=1309358
webkit: https://bugs.webkit.org/show_bug.cgi?id=165508

chromium's tracker for this was closed last december as Fixed. and i can testify that chrome works well with Access-Control-Allow-Headers: *

therefore, we need to change the Access-Control-Allow-Headers behavior away from a wildcard to a enumerated list as per https://stackoverflow.com/a/33704645

### PR DESCRIPTION
**Pull request by starpit**
As summarized in #4045, the previous #4010 PR, which changed Access-Control-Allow-Headers from a 2-element list to a wildcard was misguided. Despite whatwg introducing wildcard support in 2016, only Chrome as of yet supports this feature. So, for now, we need to back off to a specifically enumerated list of allowed headers. This PR reverts the wildcard change, but expands the list from the initial 2 to include a few more common headers, e.g. User-Agent.

Fixes #4045

## Related issue and scope

- [x] I opened an issue to propose and discuss this change (#4045)

## My changes affect the following components

- [ ] API
- [x] Controller
- [ ] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [ ] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [x] Tests
- [ ] Deployment
- [ ] CLI
- [ ] General tooling
- [x] Documentation

## Types of changes

- [x] Bug fix (generally a non-breaking change which closes an issue).
- [ ] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:

- [x] I signed an [Apache CLA](https://github.com/apache/incubator-openwhisk/blob/master/CONTRIBUTING.md).
- [x] I reviewed the [style guides](https://github.com/apache/incubator-openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [ ] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [x] I updated the documentation where necessary.

### CODE REVIEW

File: core/controller/src/main/scala/whisk/core/controller/RestAPIs.scala [hunk 1]
```diff
@@ -151,7 +151,13 @@ protected[controller] object RestApiCommons {
...
 protected[controller] trait RespondWithHeaders extends Directives {
   val allowOrigin = `Access-Control-Allow-Origin`.*
-  val allowHeaders = `Access-Control-Allow-Headers`("*")
+  val allowHeaders = `Access-Control-Allow-Headers`(
+    "Authorization",
+    "Origin",
+    "X-Requested-With",
+    "Content-Type",
+    "Accept",
+    "User-Agent")
```
Comment (markusthoemmes) [inline core/controller/src/main/scala/whisk/core/controller/RestAPIs.scala]: Should we make this list configurable?

Comment (starpit) [reply core/controller/src/main/scala/whisk/core/controller/RestAPIs.scala]: probably the ultimate right course of action is to make both fields (allow-origin, allow-headers) configurable.

the reason i've put in this short-term fix is that my previous PR probably broke some use cases; whereas previously web actions would work from any browser (assuming only those two previously allowed headers were passed), now web actions will only work from chrome (but will support *any* headers)

with this PR, we restore functionality on all browsers for web actions, with a somewhat expanded list of allow headers)

i would suggest that we do the right thing in a subsequent PR, but this PR i feel fills an important a short-term regression fix until we can get the right PR in place?

Comment (markusthoemmes) [reply core/controller/src/main/scala/whisk/core/controller/RestAPIs.scala]: SGTM

Comment (markusthoemmes) [reply core/controller/src/main/scala/whisk/core/controller/RestAPIs.scala]: Can we at least define the list of headers in a centralized place so we don't end up copying them around?

Comment (starpit) [reply core/controller/src/main/scala/whisk/core/controller/RestAPIs.scala]: sounds good. i'll work on this in a bit.

**Comment by starpit** [comment]
i have updated the PR to introduce a new file CorsSettings, and cleaned up RestAPIs and WebActions to pull the constants from this file.

in a subsequent PR, perhaps we can allow CorsSettings to be populated from the ansible/whatever level of deploy-time configuration.

**Comment by starpit** [comment]
further refined to use traits

**Comment by starpit** [comment]
ping

**Pull request merged**